core/src/main/java/org/jruby/util/io/FilenoUtil.java

```diff
@@ -126,6 +132,43 @@ public static int filenoFrom(FileDescriptor fd) {
         return -1;
     }
 
+    public static HANDLE handleFrom(Channel channel) {
+        if (channel instanceof NativeSelectableChannel) {
+            return HANDLE.valueOf(((NativeSelectableChannel)channel).getFD()); // TODO: this is an int. Do windows handles ever grow larger?
+        }
+
+        return getHandleUsingReflection(channel);
+    }
+
+    private static HANDLE getHandleUsingReflection(Channel channel) {
+        if (ReflectiveAccess.FILE_DESCRIPTOR_FD != null) {
+            return JavaLibCHelper.gethandle(getDescriptorFromChannel(channel));
+        }
+        return HANDLE.valueOf(-1);
+    }
+
+    public static int filenoFromHandleIn(Channel channel, int flags) {
+        if (winc == null)
+            return -1;
+        HANDLE hndl = handleFrom(channel);
+        if (!hndl.isValid())
+            return -1;
+        return winc._open_osfhandle(hndl.toPointer(), flags); // TODO: don't re-open this handle ever again, or we start to leak?
+    }
+
+    public static int closeFilenoHandle(int fd) {
+        if (fd != -1)
+            return winc._close(fd);// TODO: error handling
+        return -1;
+    }
+
+    public static interface WinC {
+        int _open_osfhandle(Pointer hndl, int flgs);
+        int _close(int fd);
+    }
+
+    private static final WinC winc = Platform.IS_WINDOWS ? LibraryLoader.create(WinC.class).load("msvcrt") : null;// TODO: string
```

Comment:
I think this might blow up if native support is diabled, yes? I think we need to add a check for that here and anywhere that tries to use this library.

Reply:
Pulled into ctor and accessed in a non-static manner now